src/llms/llm_loader.py

```diff
@@ -0,0 +1,233 @@
+# workaround to disable UserWarning
+import warnings
+warnings.simplefilter("ignore", UserWarning)
+
+from langchain.callbacks.streaming_stdout import StreamingStdOutCallbackHandler
+from langchain.callbacks.manager import CallbackManager
+
+import os, inspect
+
+from utils.logger import Logger
+
+class LLMLoader:
+    """
+    Note: This class loads the LLM backend libraries if the specific LLM is loaded.
+    Known caveats: Currently supports a single instance/model per backend
+
+    The llm_backends parameter is an array with the backend llm and (optional) parameters to override
+
+    # format for dictionary
+    llm_backend = {
+        'backend': 'openai',
+        'params': {...}     # optional key
+    }
+
+    # overriding specific parameters
+    llm_backend = {
+        'backend' : 'ollama',
+        'params': {'temperature': 0.02, 'top_p': 0.95},
+    }
+
+    # using default parameters
+    llm_backend = {'backend': 'tgi'}
+
+    # Using the class
+    llm_config = LLMLoader(llm_backend=llm_backend)
+    llm_chain = LLMChain(llm=llm_config.llm, prompt=prompt)
+
+    """
+    def __init__(self, llm_backend={}, logger=None) -> None:
+        self.logger = logger if logger is not None else Logger("llm_loader").logger
+        # LLM backend to use or fallback to default one
+        self.llm_backend = llm_backend.get('backend') if llm_backend.get('backend') is not None else os.environ.get('LLM_DEFAULT', {'backend':'ollama'})
+        self.llm_params = llm_backend.setdefault('params',{}) # return empty dictionary if not present
+        self.llm = None
+        self._set_llm_instance()
+
+    def _set_llm_instance(self):
+        self.logger.debug(f"[{inspect.stack()[0][3]}] Loading LLM instances with default {self.llm_backend}")
+        match self.llm_backend:
+            case 'openai':
```

Comment:
The purpose is to support multiple keys to avoid chaining the class to support other tunings in the future. The backend and params keys are the first two relevant today. Because of the difference in how params are split in libraries from IBM, vs other LLMs, it is the simplest way I found to account for variability and difference in attributes. The minimum required key is 'backend'.

Reply:
I understand the usecase. If "backend" key is a common param key to all llms, consider adding it further down the road. Eg. where you create params. I've raised this because it seems the current code can't work as you are matching dict, but casing actual value. Maybe I'm missing something, but it should be `match self.llm_backend["backend"]:`.

Reply:
self.llm_backend is not a dict, at least not alway.  The init arg is a dict, but the self.llm_backend is set here:

```
self.llm_backend = llm_backend.get('backend') if llm_backend.get('backend') is not None else os.environ.get('LLM_DEFAULT', {'backend':'ollama'})
```

which means it's either the value of the "backend" key in the `llm_backend` dict parameter argument, or it's `{'backend':'ollama'}` or whatever value the env var had.  And whether that's going to get treated by python as a dict or a string I do not know.

Clearly not good, though :)

Reply:
I will split the backend (string) and params (dict) for clarity and if no LLM definition is found, will fallback to None instead of ollama

Reply:
Updated